test/prototype/test_parq.py

```diff
@@ -53,52 +93,179 @@ def forward(self, x):
+        for n, module in model.named_modules():
+            if not _is_linear(module):
+                continue
+
+            p_orig = getattr(orig_model, n).weight  # PARQ weight
+            p = module.weight.dequantize()  # PARQ weight after quantize_
+            p_ref = getattr(m_ref, n).weight.dequantize()  # native quantize_
+
+            self.assertTrue(p_orig.equal(p_ref))
+            self.assertTrue(p.equal(p_ref))
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_4, "Test only enabled for 2.4+")
+    @common_utils.parametrize("group_size", [32, 256])
+    def test_int4_weight_only(self, group_size: int = 32):
+        model = M(m=512, n=512).to(torch.bfloat16).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model).eval().to(_DEVICE)
+        config = int4_weight_only(group_size=group_size)
+        if check_cpu_version(_DEVICE):
+            config.layout = Int4CPULayout()
+        quantize_(m_ref, config)
+
+        b = 4
+        self.compare_quantized_models(
+            model, m_ref, Int4UnifTorchaoQuantizer(), b, group_size
+        )
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    @common_utils.parametrize("group_size", [32, 512])
+    def test_intx_weight_only(self, b: int = 2, group_size: int = 32):
+        model = M(m=512, n=512).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model).eval().to(_DEVICE)
+        quantize_(
+            m_ref,
+            IntxWeightOnlyConfig(
+                weight_dtype=_BIT_WIDTH_TO_DTYPE[b], granularity=PerGroup(group_size)
+            ),
+        )
+
+        quantizer = UnifTorchaoQuantizer()
+        self.compare_quantized_models(model, m_ref, quantizer, b, group_size)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_4, "Test only enabled for 2.4+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
+    def test_int4_weight_only_e2e(self, group_size: int = 32):
+        model = M(m=512, n=512).to(torch.bfloat16).to(_DEVICE)
+        model.reset_parameters()
+
+        m_ref = copy.deepcopy(model).eval().to(_DEVICE)
+        config = int4_weight_only(group_size=group_size)
+        if check_cpu_version(_DEVICE):
+            config.layout = Int4CPULayout()
+        quantize_(m_ref, config)
+
+        b = 4
+        base_optimizer = torch.optim.AdamW(build_param_groups(model, b, group_size))
+        optimizer = QuantOptimizer(
+            base_optimizer,
+            Int4UnifTorchaoQuantizer(),
+            ProxHardQuant(),
+            quant_per_channel=True,
+        )
+        self.compare_parq_convert(model, m_ref, optimizer, config)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    def test_intx_weight_only_e2e(self, b: int = 2, group_size: int = 32):
```

Comment:
@metascroy Thanks for looking it over! I've added this end-to-end test, along with `mapping_type=MappingType.SYMMETRIC` and `target_dtype=torch.int8` defaults for `UnifTorchaoQuantizer`